lib/Sema/TypeCheckDecl.cpp

```diff
@@ -5581,6 +5583,152 @@ static Optional<ObjCReason> shouldMarkClassAsObjC(TypeChecker &TC,
   return None;
 }
 
+bool checkDeclCircularityImpl(Decl *decl, std::set<TypeBase *> &known);
+
+// Given a type, check if its in 'known'. Do the same with its subtypes.
+// When the type is nominal, get its declaration and check if any if its
+// member's type is in known by calling 'checkDeclCircularityImpl'
+bool isTypePresentForCircularity(Type type, std::set<TypeBase *> &known) {
+  auto desugared = type->getDesugaredType();
+
+  if (auto NT = dyn_cast<NominalType>(desugared)) {
+    if(auto NTT = NT->getDecl()->getDeclaredTypeInContext()) {
+      if (known.find(NTT.getPointer()) != known.end()) {
+        return true;
+      }
+    }
+    // the type is cleared, now we check its member.
+    return checkDeclCircularityImpl(NT->getDecl(), known);
+  }
+
+  // check each generic argument, this takes care of parameter from declaring
+  // context as well as stuff like Optionals
+  if (auto BGT = dyn_cast<BoundGenericType>(desugared)) {
+    for (auto T: BGT->getGenericArgs()) {
+      if(isTypePresentForCircularity(T, known)) {
+        return true;
+      }
+    }
+
+    // check the overall generic type
+    auto BGTT = BGT->getDecl()->getDeclaredTypeInContext();
+    if (known.find(BGTT.getPointer()) != known.end()) {
+      return true;
+    }
+
+    // check this generic type's members.
+    return checkDeclCircularityImpl(BGT->getDecl(), known);
+  }
+
+  // if this is a tuple, deconstruct it and check each element
+  if (auto tuple = dyn_cast<TupleType>(desugared)) {
+    for (auto Elt: tuple->getElements()) {
+      if (isTypePresentForCircularity(Elt.getType(), known)) {
+        return true;
+      }
+    }
+  }
+
+  if (known.find(desugared) != known.end()) {
+    return true;
+  }
+  return false;
+}
+
+
+// Given a declaration, iterate over its members and make sure non of them
+// is of a value type of the very same declaration. Use
+// 'isTypePresentForCircularity' to exam related types of each member.
+bool checkDeclCircularityImpl(Decl *decl, std::set<TypeBase *> &known) {
```

Comment:
Also I think you should use DenseSet or SmallPtrSet here, since you don't need ordering